src/tools/rbd/action/MirrorSnapshotSchedule.cc

```diff
@@ -0,0 +1,607 @@
+
+namespace {
+
+struct Args {
+  std::map<std::string, std::string> args;
+
+  Args() {
+  }
+  Args(const std::map<std::string, std::string> &args)
+    : args(args) {
+  }
+
+  std::string str() const {
+    std::string out = "";
+
+    std::string delimiter;
+    for (auto &it : args) {
+      out += delimiter + "\"" + it.first + "\": \"" +
+        stringify(json_stream_escaper(it.second)) + "\"";
+      delimiter = ",\n";
+    }
+
+    return out;
+  }
+};
+
+class Schedule {
+public:
+  Schedule() {
+  }
+
+  int parse(const std::string &schedule) {
+    json_spirit::mValue json_root;
+    if(!json_spirit::read(schedule, json_root)) {
+          std::cerr << "rbd: invalid schedule JSON received" << std::endl;
+      return -EBADMSG;
+    }
+
+    return parse(json_root);
+  }
+
+  int parse(json_spirit::mValue &schedule_val) {
+    if (schedule_val.type() != json_spirit::array_type) {
+      std::cerr << "rbd: unexpected schedule JSON received" << std::endl;
+      return -EBADMSG;
+    }
+
+    try {
+      for (auto &item_val : schedule_val.get_array()) {
+        if (item_val.type() != json_spirit::obj_type) {
+          std::cerr << "rbd: unexpected schedule JSON received" << std::endl;
+          return -EBADMSG;
+        }
+
+        auto &item = item_val.get_obj();
+
+        if (item["interval"].type() != json_spirit::str_type) {
+          std::cerr << "rbd: unexpected schedule JSON received" << std::endl;
+          return -EBADMSG;
+        }
+        auto interval = item["interval"].get_str();
+
+        std::string start_time;
+        if (item["start_time"].type() == json_spirit::str_type) {
+          start_time = item["start_time"].get_str();
+        }
+
+        items.push_back({interval, start_time});
+      }
+
+    } catch (std::runtime_error &) {
+      std::cerr << "rbd: unexpected schedule JSON received" << std::endl;
+      return -EBADMSG;
+    }
+
+    return 0;
+  }
+
+  void dump(Formatter *f) {
+    f->open_array_section("items");
+    for (auto &item : items) {
+      f->open_object_section("item");
+      f->dump_string("interval", item.first);
+      f->dump_string("start_time", item.second);
+      f->close_section(); // item
+    }
+    f->close_section(); // items
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, Schedule &s);
+
+private:
+  std::string name;
+  std::list<std::pair<std::string, std::string>> items;
+};
+
+std::ostream& operator<<(std::ostream& os, Schedule &s) {
+  std::string delimiter;
+  for (auto &item : s.items) {
+    os << delimiter << "every " << item.first;
+    if (!item.second.empty()) {
+      os << " starting at " << item.second;
+    }
+    delimiter = ", ";
+  }
+  return os;
+}
+
+int parse_schedule_name(const std::string &name, std::string *pool_name,
+                        std::string *namespace_name, std::string *image_name) {
+  // parse names like:
+  // '', 'rbd/', 'rbd/ns/', 'rbd//image', 'rbd/image', 'rbd/ns/image'
```

Comment:
Nit: `rbd//image` wouldn't be valid under the new scheme, correct?

Reply:
Yes, it will always come from the scheduler module as `rbd/image`, though the scheduler module "understands" both ways. It does not hurt but ok I will remove this redundancy here.